### ISSUE TITLE
Preprocessing Steps

### ISSUE THREAD

**Issue by likith012** [dthiagarajan/ptb-xl-research#19]
Can you explain what processing steps are you following for the ptb-xl dataset.

**Comment by dthiagarajan** [comment]
Almost all of the processing is handled with code in `core/data/`. In particular:

- Data is loaded in as specified in the utility given by the dataset authors: [link](https://github.com/dthiagarajan/ptb-xl-research/blob/185a8ccb9c61296d7995069c5f16d652805ae908/core/data/utils.py#L77)
- Data is split using the split numbers given in the dataset: [link](https://github.com/dthiagarajan/ptb-xl-research/blob/185a8ccb9c61296d7995069c5f16d652805ae908/core/data/utils.py#L121)
- Labels are binarized using `sklearn`: [link](https://github.com/dthiagarajan/ptb-xl-research/blob/185a8ccb9c61296d7995069c5f16d652805ae908/core/data/transforms.py#L86)

Then, during training, a random window is sampled to train on: [link](https://github.com/dthiagarajan/ptb-xl-research/blob/185a8ccb9c61296d7995069c5f16d652805ae908/core/data/transforms.py#L19).

During validation or testing, all windows (in order) of the same size as in training are taken, and the prediction over them is averaged: [link](https://github.com/dthiagarajan/ptb-xl-research/blob/185a8ccb9c61296d7995069c5f16d652805ae908/core/data/transforms.py#L35).

Let me know if this clarifies your question.

**Comment by likith012** [comment]
I see that the PTB-xl dataset is a heavily imbalanced dataset, are there any consistent synthetic data generation techniques are you aware of, especially considering these are 1D signals.

**Comment by dthiagarajan** [comment]
Synthetic data generation / data augmentation isn't that beneficial for heavily imbalanced datasets in my experience, but one technique I use here is [mixup](https://github.com/dthiagarajan/ptb-xl-research/blob/master/core/callbacks/model_callbacks.py#L138). This scheme of mixup / manifold mixup could be augmented slightly to prioritize mixing more of examples that belong to sparse classes.

Another technique I use, unrelated to data augmentation, is class-weighted loss, which will emphasize examples from sparse classes more due to their infrequency. That implementation can be found [here](https://github.com/dthiagarajan/ptb-xl-research/blob/master/core/models/loss_functions.py#L25).

**Comment by likith012** [comment]
Can this approach [https://arxiv.org/abs/2002.02870#:~:text=version%2C%20v2)%5D-,The%20Effect%20of%20Data%20Augmentation%20on%20Classification%20of%20Atrial%20Fibrillation,Signals%20Using%20Deep%20Neural%20Networks&text=The%20results%20show%20that%20deep,GANs%20and%20GMMs%2C%20than%20oversampling.](url) be used for this particular dataset as they were able to obtain performance improvements by synthetic data generation. Any thoughts on this from your end?

**Comment by dthiagarajan** [comment]
Possibly? Though I think it'd be more reasonable to start with a simpler data augmentation scheme and explore adding complexity to that rather than using something more complex to start.

**Comment by likith012** [comment]
I see that you are using 2.5 sec of ECG signals for modelling, any particular reason you considering 2.5 sec and why aren't you considering the complete 10 sec ECG signal for modelling. Thanks in advance.

**Comment by dthiagarajan** [comment]
This is to conform with the reference paper: "If not mentioned otherwise, we use a fixed window size of 2.5 seconds."

https://arxiv.org/pdf/2004.13701.pdf

**Comment by likith012** [comment]
Hey got a doubt regarding the metrics,  the authors in the benchmarking paper have used the sklearn's `auc_roc` with macro averaged, does the prediction vector input to this along with ground truth should be of the form `[1, 1, 0, 0, 1]`(rounded off values) or `[0.78, 0.89, 0.4, 0.1, .67]`(probability scores), right now I am confused as both are giving different results, I thought both will be give the same `auc_roc` score.

**Comment by likith012** [comment]
I believe you are using the same metrics as in this repo https://github.com/helme/ecg_ptbxl_benchmarking/blob/dd30e4940b12140fef40aeddaf21a7af7c565a1e/code/utils/utils.py#L23

can you confirm whether the `y_pred` is in rounded off form `[1, 1, 0, 0, 1]` as opposed to probability values  in function call `evaluate_experiments`

**Comment by dthiagarajan** [comment]
You can find my logic here: https://github.com/dthiagarajan/ptb-xl-research/blob/master/core/metrics.py

**Comment by dthiagarajan** [comment]
In particular, I use probabilities, via [this function](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.roc_auc_score.html), on a per-label basis, then take the macro-average (mean).

**Comment by dthiagarajan** [comment]
Thanks for your questions - I'm going to close this issue, but please feel free to open a new one specific to any further questions you have.

**Comment by likith012** [comment]
> Almost all of the processing is handled with code in `core/data/`. In particular:
> 
> * Data is loaded in as specified in the utility given by the dataset authors: [link](https://github.com/dthiagarajan/ptb-xl-research/blob/185a8ccb9c61296d7995069c5f16d652805ae908/core/data/utils.py#L77)
> * Data is split using the split numbers given in the dataset: [link](https://github.com/dthiagarajan/ptb-xl-research/blob/185a8ccb9c61296d7995069c5f16d652805ae908/core/data/utils.py#L121)
> * Labels are binarized using `sklearn`: [link](https://github.com/dthiagarajan/ptb-xl-research/blob/185a8ccb9c61296d7995069c5f16d652805ae908/core/data/transforms.py#L86)
> 
> Then, during training, a random window is sampled to train on: [link](https://github.com/dthiagarajan/ptb-xl-research/blob/185a8ccb9c61296d7995069c5f16d652805ae908/core/data/transforms.py#L19).
> 
> During validation or testing, all windows (in order) of the same size as in training are taken, and the prediction over them is averaged: [link](https://github.com/dthiagarajan/ptb-xl-research/blob/185a8ccb9c61296d7995069c5f16d652805ae908/core/data/transforms.py#L35).
> 
> Let me know if this clarifies your question.

Hey, sorry to bother you again, I just had one more doubt.

During testing, I used the sliding technique to segment the windows using half the overlap as you have done.
When you meant `averaged the prediction over them` did you mean averaging the `prediction probabilities` for each sample i.e, averaging the prediction  probabilities of all the windowed segments in one sample.

**Comment by dthiagarajan** [comment]
Yep, exactly!

**Comment by likith012** [comment]
Hey, have you used any cross-validation techniques(k fold or sk fold) for the models.

**Comment by dthiagarajan** [comment]
Not explicitly, but it's definitely possible (cross-validating across window size, hyperparameters of other modifications such as which loss function, parameters of the specific loss function, etc.)

**Comment by likith012** [comment]
Hey, I see that your best models are getting a roc_auc score of around 0.9235, whatever I do I wasn't able to cross the 0.9252 score (tried implementing combinations of spectrograms too), I wonder how the authors in the ptb xl benchmarking are able to get a roc score of 0.93(super diagnostic classes), by any chance do you have any explanation for this.

**Comment by dthiagarajan** [comment]
I don't have an initial guess as to why that's happening - I tried a number of things they did, like matching up hyper-parameters, using the same loss function and learning rate schedule, and so on. I would imagine it might come down to architectural differences, and how they initialize weights. I would read what they did more closely re: architecture to learn more, but my focus here wasn't necessarily to achieve SotA, rather to get heatmaps that would be viable in clinical workflows.

On a side note, please open up a new issue should you have questions related to different topics (e.g. this would be under an issue for performance, perhaps) - I'd like for future users to have an easier time searching these discussions. Thanks!
